src/Stratis.Bitcoin/Utilities/MemoryCache.cs

```diff
@@ -0,0 +1,135 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    public class MemoryCache<TKey, TValue>
+    {
+        /// <summary>Dictionary that contains cached items.</summary>
+        private readonly ConcurrentDictionary<TKey, TValue> cache;
+
+        /// <summary>Keys sorted by time they when they were accessed last time with most recent ones in the end.</summary>
+        private readonly LinkedList<TKey> keys;
+
+        /// <summary>Maximum items count that can be stored in the cache.</summary>
+        private readonly int sizeLimit;
+
+        /// <summary>Amount to compact the cache by when the maximum size is exceeded.</summary>
+        private readonly double compactionPercentage;
+
+        /// <summary>Lock to protect access to <see cref="keys"/>.</summary>
+        private readonly object mutex;
+        
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MemoryCache{TKey, TValue}"/> class.
+        /// </summary>
+        /// <param name="sizeLimit">Maximum items count that can be stored in the cache.</param>
+        /// <param name="compactionPercentage">Amount to compact the cache by when the maximum size is exceeded.</param>
+        public MemoryCache(int sizeLimit, double compactionPercentage)
+        {
+            this.sizeLimit = sizeLimit;
+            this.compactionPercentage = compactionPercentage;
+
+            this.cache = new ConcurrentDictionary<TKey, TValue>();
+            this.keys = new LinkedList<TKey>();
+            this.mutex = new object();
+        }
+
+        /// <summary>Gets the count of the current entries for diagnostic purposes.</summary>
+        public int Count
+        {
+            get
+            {
+                lock (this.mutex)
+                {
+                    return this.keys.Count;
+                }
+            }
+        }
+
+        /// <summary>Create or overwrite an entry in the cache.</summary>
+        /// <param name="key">The key.</param>
+        /// <param name="value">The value to add to the cache.</param>
+        public void AddOrUpdate(TKey key, TValue value)
+        {
+            if (this.cache.TryGetValue(key, out TValue priorEntry))
+            {
+                this.cache.TryUpdate(key, value, priorEntry);
+
+                lock (this.mutex)
+                {
+                    this.keys.Remove(key);
+                    this.keys.AddLast(key);
+                }
+            }
+            else
+            {
+                this.cache.TryAdd(key, value);
+
+                lock (this.mutex)
+                {
+                    this.keys.AddLast(key);
+                }
+
+                if (this.Count >= this.sizeLimit)
```

Comment:
you should keep your invariants - i.e. there should never be more than max number of items in the cache
you should first compact the structure and only add then

Reply:
Fixed.